denonavr/api.py

```diff
@@ -309,3 +313,184 @@ def prepare_appcommand_body(cmd_list: Tuple[AppCommandCmd]) -> bytes:
+                await self._socket_writer.wait_closed()
+            self._socket_writer = None
+
+    async def _async_monitor(self):
+        """Reads the messages on the TCP socket."""
+        data = bytearray()
+        while not self._socket_reader.at_eof():
+            try:
+                chunk = await asyncio.wait_for(self._socket_reader.read(_SOCKET_READ_SIZE), self.timeout)
+                for i in range(0,len(chunk)):
+                    # Messages are CR terminated
+                    if chunk[i] != 13:
+                        data += chunk[i].to_bytes(1, byteorder='big')
+                    else:
+                        await self._process_event(str(data,'utf-8'))
+                        data = bytearray()
+            except asyncio.exceptions.TimeoutError as err:
+                _LOGGER.debug("Lost connection to receiver, reconnecting")
+                await self.async_disconnect()
+                await self.async_connect()
+
+    def register_callback(self, type: str="ALL", callback=lambda *args: any):
+        """Registers a callback handler for an event type."""       
+        if not type in self._callbacks.keys():
+            self._callbacks[type] = []
+        self._callbacks[type].append(callback)
+
+    def unregister_callback(self, type: str="ALL", callback=lambda *args: any):
+        """Unregisters a callback handler for an event type."""
+        if not type in self._callbacks.keys():
+            return
+        self._callbacks[type].remove(callback)
+
+    async def _process_event(self, message):
+        """Process a realtime event."""
+        if len(message) < 3:
+            return None
+        
+        #Event is 2 characters
+        event = message[0:2]
+        #Parameter is the remaining characters
+        parameter = message[2:]
+        
+        if event == 'PW':
+            await self._process_power(MAIN_ZONE, parameter)
+        elif event == 'MV':
+            await self._process_volume(MAIN_ZONE, parameter)
+        elif event == 'MU':
+            await self._process_mute(MAIN_ZONE, parameter)
+        elif event == 'SI':
+            await self._process_input(MAIN_ZONE, parameter)
+        elif event == 'MS':
+            await self._process_surroundmode(MAIN_ZONE, parameter)
+        elif event == 'PS':
+            await self._process_sounddetail(MAIN_ZONE, parameter)
+        
+        elif event == 'Z2' or event == 'Z3':
+            if event == 'Z2':
+                zone = ZONE2
+            else:
+                zone = ZONE3
+            if parameter == 'ON' or parameter == 'OFF':
+                await self._process_power(zone, parameter)
+            elif parameter == 'MUON' or parameter == 'MUOFF':
+                await self._process_mute(zone, parameter)
+            elif parameter in ZONE_SOURCES:
+                await self._process_input(zone, parameter)
+            elif parameter.isdigit():
+                await self._process_volume(zone, parameter)        
```

Comment:
I can do that, the problem is it seemed like this would be a poor separation of concerns since the individual class (e.g. mute) has to know how to parse logic about the overarching API message. Meaning the mute class now becomes tightly coupled.

Reply:
I went with a compromise here that I hope you'll agree makes sense. I believe the Z2 vs Z3 handling is a detail of the overall API and should not be handled internally by each command. Meaning, I believe that is an implementation detail of the api not of volume vs mute vs etc. So I think we should leave this in api.py but other than that exception, now all command processing logic is in the individual classes. What do you think?